azure-iot-device/tests/common/pipeline_test.py

```diff
@@ -0,0 +1,117 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+import pytest
+import functools
+from azure.iot.device.common.transport import pipeline_events_base
+from azure.iot.device.common.transport import pipeline_ops_base
+from azure.iot.device.common.transport import pipeline_stages_base
+from azure.iot.device.common.transport.mqtt import pipeline_events_mqtt
+from azure.iot.device.common.transport.mqtt import pipeline_ops_mqtt
+
+all_common_ops = [
+    [pipeline_ops_base.Connect, []],
+    [pipeline_ops_base.Reconnect, []],
+    [pipeline_ops_base.Disconnect, []],
+    [pipeline_ops_base.EnableFeature, [""]],
+    [pipeline_ops_base.DisableFeature, [""]],
+    [pipeline_ops_base.SetSasToken, [""]],
+    [pipeline_ops_mqtt.SetConnectionArgs, ["", "", ""]],
+    [pipeline_ops_mqtt.Publish, ["", ""]],
+    [pipeline_ops_mqtt.Subscribe, [""]],
+    [pipeline_ops_mqtt.Unsubscribe, [""]],
+]
+
+all_common_events = [
+    [pipeline_events_base.PipelineEvent, []],
+    [pipeline_events_mqtt.IncomingMessage, ["", ""]],
+]
+
+
+def all_except(all_items, items_to_exclude):
+    """
+    helper function to return a new list with all ops that are in the first list
+    and not in the second list.
+
+    :param list all_items: list of all operations or events
+    :param list items_to_exclude: ops or events to exclude
+    """
+    return [x for x in all_items if x[0] not in items_to_exclude]
+
+
+def assert_default_stage_attributes(obj):
+    assert obj.name is obj.__class__.__name__
+    assert obj.next is None
+    assert obj.previous is None
+    assert obj.pipeline_root is None
+
+
+# because PipelineStage is abstract, we need something concrete
+class ConcretePipelineStage(pipeline_stages_base.PipelineStage):
+    def _run_op(self, op):
+        self.continue_op(op)
```

Comment:
Are we still doing this?

Reply:
I think you're talking about  not testing abstract classes.  I want to cover that under a different PR.  

---
In reply to: [287540616](https://github.com/Azure/azure-iot-sdk-python-preview/pull/94#discussion_r287540616) [](ancestors = 287540616)